### PR TITLE
[KernelGen][Nvidia] Add permute_copy operator

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
- Add `permute_copy` operator implemented with Triton kernels for 1D, 2D, and 3D tensors
- Ranks > 3 fall back to permute + contiguous for correctness
- Includes accuracy tests and performance benchmark

## Performance (Nvidia)
| Metric | Value |
|--------|-------|
| Cases | 15 |
| Mean Speedup | 1.026x |

## Multi-platform Results
| Platform | Accuracy | Speedup |
|----------|----------|---------|
| Nvidia | PASS | 1.026x |
| Tianshu | FAIL | 0.716x |
| Muxi | PASS | 0.882x |
| Ascend | PASS | — |
| Hygon | PASS | 1.056x |

## Test plan
- [x] Local pytest accuracy tests pass
- [x] Local benchmark runs successfully
- [x] check_operator.py --strict passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: tests/test_permute_copy.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    ((2, 3), (1, 0)),
+    ((3, 4), (1, 0)),
+    ((16, 128), (1, 0)),
+    # 3D cases (from original PERMUTE_COPY_SHAPES x PERMUTE_COPY_PERMS)
+    ((2, 3, 4), (0, 2, 1)),
+    ((2, 3, 4), (1, 0, 2)),
+    ((2, 3, 4), (2, 1, 0)),
+    ((4, 2, 3), (1, 2, 0)),
+    ((4, 2, 3), (2, 0, 1)),
+]
```
Comment (bin913) [inline tests/test_permute_copy.py:17]: Please centrally manage these test shapes in accuracy_utils.py

**Comment by XDYuanzhuLee** [comment]
Addressed in the latest push: moved PERMUTE_COPY_SHAPES_AND_DIMS to tests/accuracy_utils.py for centralised management.

File: tests/test_permute_copy.py (hunk 1 above)
Comment (XDYuanzhuLee) [reply tests/test_permute_copy.py:17]: Done. `PERMUTE_COPY_SHAPES_AND_DIMS` is now centrally managed in `tests/accuracy_utils.py`, and the test uses the shared constant.

**Comment by bin913** [comment]
@XDYuanzhuLee Please fix these failing checks

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator 贡献审查报告

**Operator**: `permute_copy`
**PR**: #3535

### 📊 变更统计
- 新增行数: 319
- 删除行数: 0  
- 变更文件: 6

### ✅ 检查清单

| 检查项 | 状态 |
|--------|------|
| 实现文件 (`src/flag_gems/ops/permute_copy.py`) | ✅ 存在 |
| 测试文件 (`tests/test_permute_copy.py`) | ✅ 存在 |
| Benchmark 文件 | ✅ 存在 |
| 配置注册 (`operators.yaml`) | ✅ 存在 |
| 模块注册 (`__init__.py`) | ✅ 存在 |
| KernelGen 标注 | ✅ 存在 |
| Triton JIT 装饰器 | ✅ 存在 |
| Metax Backend | ⚠️ 未实现 |

### 📝 审查意见

**代码结构**: 实现文件结构正确

**测试覆盖**: 有测试文件

**性能测试**: 有 benchmark

---

*此审查由自动化脚本根据 FlagGems Operator 贡献规范生成。*
*如有问题请联系维护者进行详细审查。*

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `permute_copy`  
**PR**: #3535

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 319 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/permute_copy.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ⚠️ `@triton.jit` present, verify other decorators |
| 7 | Test file existence | ✅ Present: `tests/test_permute_copy.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_permute_copy.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ✅ Stability patterns found |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ✅ Docstrings present |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 13 |
| ⚠️ Warning | 8 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Comment by XDYuanzhuLee** [comment]
XDYuanzhuLee has fixed the failing checks in commits e61d5651 (merge master) and 0dc29d63 (fix: isort import style and OpsConfigFile NameError). All CI checks are now passing:
- code-style: pass
- triage: pass
- python-op: pass
- preprocess: pass

**Review by Dongxu-H — APPROVED** [review]
## Code Review: permute_copy Implementation

Thanks for implementing the `permute_copy` operator! The algorithm is correct and CI passes. However, I have some concerns about the performance benefits and code efficiency.

---

### ✅ Algorithm Correctness

The core permutation logic is correctly implemented:

```python
# Line 55-59: Compute perm_inv
# perm_inv[d] = j such that perm[j] = d
perm_inv0 = tl.where(perm1 == 0, 1, 0)
perm_inv1 = tl.where(perm0 == 1, 0, 1)

# Line 61-65: Compute source offset
# input_idx[d] = output_idx[perm_inv[d]]
src_offset = (
    tl.where(perm_inv0 == 0, o0, o1) * src_stride0
    + tl.where(perm_inv1 == 0, o0, o1) * src_stride1
)
```

**Example verification**: For `x.shape = (3, 4)`, `dims = (1, 0)`:
- Output shape = `(4, 3)`
- `perm_inv[0] = 1`, `perm_inv[1] = 0`
- `src_offset = o1 * src_stride[0] + o0 * src_stride[1]` ✅

---

### ⚠️ Performance Concerns

**Reported speedup**:

| Platform | Speedup |
|----------|---------|
| Nvidia | 1.026x |
| Tianshu | 0.716x (regression!) |
| Muxi | 0.882x (regression!) |
| Hygon | 1.056x |

**Issues**:
- Average speedup is only **1.026x** - nearly negligible
- **Tianshu and Muxi show performance regression**
- `permute_copy` is fundamentally a memory copy operation, limited by memory bandwidth

**Analysis**:
1. `permute` itself is lightweight (just stride transformation)
2. `contiguous()` is memory-intensive
3. Triton kernel launch overhead may offset benefits

---

### ⚠️ Code Efficiency Issues

#### 1. 1D Kernel is Unnecessary

```python
# Line 34-42
# Single dimension - no permutation possible
o0 = offsets
perm_inv0 = perm0
src_offset = tl.where(perm_inv0 == 0, o0, 0) * src_stride0
```

**Problem**:
- For 1D tensors, `permute` is a no-op (only one dimension)
- `dims[0]` must be 0 (only valid value)
- This kernel is equivalent to a simple `clone()`

**Suggestion**: For 1D tensors, just return `x.clone()` without launching kernel:

```python
if ndim == 1:
    return x.clone()  # No permutation possible
```

#### 2. Unnecessary Contiguous Check

```python
# Line 196-198
src = x.contiguous() if not x.is_contiguous() else x
if not out.is_contiguous():  # ← Always False!
    out = out.contiguous()
```

**Problem**: `out = torch.empty(out_shape, ...)` creates a contiguous tensor by default. The check adds unnecessary runtime overhead.

**Suggestion**: Remove the redundant check:

```python
src = x.contiguous() if not x.is_contiguous() else x
# out is already contiguous
```

---

### ⚠️ Test Coverage Gaps

**Current tests** cover main cases but missing edge cases:

- ⚠️ Empty tensor: `torch.empty(0, 3, 5)`
- ⚠️ 1D tensor: `torch.randn(10)` with `dims=(0,)`
- ⚠️ Identity permutation: `dims=(0, 1)` for 2D

**Suggested tests**:

```python
def test_permute_copy_empty():
    """Test with empty tensor."""
    inp = torch.empty(0, 3, 5)
    ref_out = torch.permute_copy(inp, (1, 0, 2))
    with flag_gems.use_gems():
        res_out = torch.permute_copy(inp, (1, 0, 2))
    assert res_out.shape == ref_out.shape

def test_permute_copy_1d():
    """Test 1D tensor (no-op case)."""
    inp = torch.randn(10)
    ref_out = torch.permute_copy(inp, (0,))
    with flag_gems.use_gems():
        res_out = torch.permute_copy(inp, (0,))
    utils.gems_assert_close(res_out, ref_out, inp.dtype)

def test_permute_copy_identity():
    """Test identity permutation."""
    inp = torch.randn(4, 8)
    ref_out = torch.permute_copy(inp, (0, 1))
    with flag_gems.use_gems():
        res_out = torch.permute_copy(inp, (0, 1))
    utils.gems_assert_close(res_out, ref_out, inp.dtype)
```

---

### ✅ What's Good

1. **Correct algorithm** - Index mapping formula is correct
2. **Safe fallback** - Uses PyTorch native for rank > 3
3. **Empty tensor handling** - `if x.numel() == 0` correctly handled
4. **Negative index support** - `dims = [d if d >= 0 else d + ndim for d in dims]`
5. **CI passes** - python-op, code-style pass
6. **Complete registration** - operators.yaml, __init__.py, ops/__init__.py

---

### 📊 Summary

| Aspect | Status |
|--------|--------|
| Algorithm | ✅ Correct |
| Performance | ⚠️ 1.026x speedup, some platforms regress |
| 1D Kernel | ⚠️ Unnecessary, should simplify |
| Code Efficiency | ⚠️ Redundant checks |
| Test Coverage | ⚠️ Missing edge cases |
| CI Status | ✅ All pass |

---

### 🎯 Conclusion

**Approved** ✅ - The implementation is algorithmically correct and ready for merge. However, I recommend considering:

1. **Simplify 1D case** - Just return `x.clone()` without kernel launch
2. **Remove redundant checks** - `out.is_contiguous()` is always True
3. **Add edge case tests** - Empty, 1D, and identity permutation cases
4. **Consider value** - With only 1.026x speedup and some regressions, evaluate if Triton kernel is worth the complexity

Great work on the implementation!

**Review by Caeruleann — APPROVED** [review]
